### ISSUE TITLE
Error when training Resnet with nn.DistributedDataParallel

### ISSUE THREAD

**Issue by jiahaosuda** [pytorch/vision#2505]
When I try to train [Resnet](https://github.com/pytorch/vision/blob/master/torchvision/models/resnet.py) with `nn.DistributedDataParallel` using multi-gpus, the error occurs as below. But when I use only one gpu, it's just ok. 

![image](https://user-images.githubusercontent.com/40142236/88374211-ed35c280-cdcb-11ea-8f8b-e452e52c11ce.png)

Some of the code:
```
class Bottleneck(nn.Module):
    expansion = 4
    def __init__(self, inplanes, planes, stride=1, downsample=None):
        super(Bottleneck, self).__init__()
        self.conv1 = nn.Conv2d(inplanes, planes, kernel_size=1, bias=False)  # decrease the channel, does't change size
        self.bn1 = nn.BatchNorm2d(planes)
        self.conv2 = nn.Conv2d(planes, planes, kernel_size=3, stride=stride,
                               padding=1, bias=False)
        self.bn2 = nn.BatchNorm2d(planes)
        self.conv3 = nn.Conv2d(planes, planes * 4, kernel_size=1, bias=False)
        self.bn3 = nn.BatchNorm2d(planes * 4)
        self.relu = nn.ReLU(inplace=False)
        self.downsample = downsample
        self.stride = stride

    def forward(self, x):
        residual = x

        out = self.conv1(x)
        out = self.bn1(out)
        out = self.relu(out)

        out = self.conv2(out)
        out = self.bn2(out)
        out = self.relu(out)

        out = self.conv3(out)
        out = self.bn3(out)

        if self.downsample is not None:
            residual = self.downsample(x)

        out = out + residual
        out = self.relu(out)

        return out


class ResNet(nn.Module):

    def __init__(self, block, layers, num_classes=9):
        self.inplanes = 64
        super(ResNet, self).__init__()
        self.conv1 = nn.Conv2d(3, 64, kernel_size=7, stride=2, padding=3,
                               bias=False)  # the size become 1/2
        self.bn1 = nn.BatchNorm2d(64)
        self.relu = nn.ReLU(inplace=False)
        self.maxpool = nn.MaxPool2d(kernel_size=3, stride=2, padding=1)  # the size become 1/2
        self.layer1 = self._make_layer(block, 64, layers[0])
        self.layer2 = self._make_layer(block, 128, layers[1], stride=2)
        self.layer3 = self._make_layer(block, 256, layers[2], stride=2)
        self.layer4 = self._make_layer(block, 512, layers[3], stride=2)
        self.avgpool = nn.AdaptiveAvgPool2d((1, 1))
        # self.fc = nn.Linear(512 * block.expansion, num_classes)
        self.fc1 = nn.Linear(2048, 1024)
        self.fc2 = nn.Linear(1024, num_classes)


        for m in self.modules():
            if isinstance(m, nn.Conv2d):
                n = m.kernel_size[0] * m.kernel_size[1] * m.out_channels
                m.weight.data.normal_(0, math.sqrt(2. / n))
            elif isinstance(m, nn.BatchNorm2d):
                m.weight.data.fill_(1)
                m.bias.data.zero_()

    def _make_layer(self, block, planes, blocks, stride=1):
        #  block: object, planes: output channel, blocks: the num of blocks
        downsample = None
        if stride != 1 or self.inplanes != planes * block.expansion:
            downsample = nn.Sequential(
                nn.Conv2d(self.inplanes, planes * block.expansion,
                          kernel_size=1, stride=stride, bias=False),
                nn.BatchNorm2d(planes * block.expansion),
            )

        layers = []
        layers.append(block(self.inplanes, planes, stride, downsample))
        self.inplanes = planes * block.expansion  # the input channel num become 4 times
        for i in range(1, blocks):
            layers.append(block(self.inplanes, planes))

        return nn.Sequential(*layers)

    def forward(self, x):
        x = self.conv1(x)
        x = self.bn1(x)
        x = self.relu(x)
        x = self.maxpool(x)

        x = self.layer1(x)
        x = self.layer2(x)
        x = self.layer3(x)
        x = self.layer4(x)

        x = self.avgpool(x)
        x = x.view(x.size(0), -1)
        x = self.fc1(x)
        x = nn.init.normal_(x, mean=0, std=1024 ** -0.5)
        # x = self.fc2(x)
        return x
```

**Comment by vfdev-5** [comment]
@jiahaosuda actually, what is the purpose of `x = nn.init.normal_(x, mean=0, std=1024 ** -0.5)` in `ResNet.forward` of your code ?

**Comment by jiahaosuda** [comment]
> @jiahaosuda actually, what is the purpose of `x = nn.init.normal_(x, mean=0, std=1024 ** -0.5)` in `ResNet.forward` of your code ?
> 
> Anyway, this is the inplace operation mentioned in the error.

By using `x = nn.init.normal_(x, mean=0, std=1024 ** -0.5)`, i just try to extract the feature of the image to act as the its embedding (as same as word embedding). How can i do instead of this usage?

But when i remove `x = nn.init.normal_(x, mean=0, std=1024 ** -0.5)` from the code, the error still exists. And the error shows something around `out = self.bn3(out)` may get wrong, so i guess maybe `nn.BatchNorm2d(num)` when `nn.DistributedDataParallel` goes wrong?

**Comment by vfdev-5** [comment]
@jiahaosuda sorry for delayed reply !

> But when i remove x = nn.init.normal_(x, mean=0, std=1024 ** -0.5) from the code, the error still exists. And the error shows something around out = self.bn3(out) may get wrong, so i guess maybe nn.BatchNorm2d(num) when nn.DistributedDataParallel goes wrong?

Normally, DDP training of the origin resnet should work without problems : https://github.com/pytorch/examples/tree/master/imagenet
Without any reproducible code it would be rather difficult to inspect the problem of your ResNet implementation...

> By using x = nn.init.normal_(x, mean=0, std=1024 ** -0.5), i just try to extract the feature of the image to act as the its embedding (as same as word embedding). How can i do instead of this usage?

`nn.init.normal_` is an operation that fills the input Tensor with values drawn from the normal distribution: https://pytorch.org/docs/stable/nn.init.html#torch.nn.init.normal_ so it replaces the feature maps produced by `x = self.fc1(x)` by a normal distribution. Is that what you would like to do ?

**Comment by fmassa** [comment]
Hi,

The error you are facing is what @vfdev-5 mentioned: you probably don't want to do an in-place operation in your tensor in the forward, as you'll lose the data you had in there.

The other error you are facing is probably because now `fc2` is not used anymore, and as such you are facing an error such as "some parameters are unused" in DDP, which you can fix by either removing the `fc2` from the init, or by setting `find_unused_parameters=True` in DDP.

I believe we have answered your issues, and as such I'm closing this issue but let us know if you have further problems

**Comment by jiahaosuda** [comment]
@vfdev-5 @fmassa Sorry for late reply. You're right. I solved the problems you mentioned and it works now. Thanks.
